### ISSUE TITLE
Use stdin automatically

### ISSUE THREAD

**Issue by bfred-it** [sindresorhus/xo#91]
Just a thought, considering that other tools just detect it and support it directly, but it might be a breaking change.

**Comment by sindresorhus** [comment]
Don't really see how we can do that as `$ xo` is run without arguments by default, and assuming stdin when `process.stdin.isTTY` means there's no way to execute it as a child process (`childProcess.execSync('xo')`). Open to suggestions though.

**Comment by bfred-it** [comment]
I see what you mean now (had to Google how [TTY and stdin were connected](http://stackoverflow.com/a/15485424/288906)). It sounds strange to me that there's no way to tell the two situations apart, but then again I'm a beginner with cli.

**Comment by sindresorhus** [comment]
> It sounds strange to me that there's no way to tell the two situations apart, but then again I'm a beginner with cli.

I couldn't agree more.

What I do in most of my CLI's is to fallback to reading stdin if there's no input.

@Qix- Is there any better way of handling this? Like being able to tell if stdin is used without having to try reading from it first?

**Comment by Qix-** [comment]
> What I do in most of my CLI's is to fallback to reading stdin if there's no inp

This is how most programs work. Take a look at `cat` for example. If there are filenames specified then it just cats them out and is done. Otherwise, it reads from stdin. It also starts reading from stdin if `-` is specified as a filename.

Why not remove `--stdin` (because it's ugly) and support `-` as a filename indicating that stdin should be read?

```console
$ xo foo.js - bar.js
read foo.js...
read stdin...
read bar.js...
```

**Comment by sindresorhus** [comment]
> Why not remove --stdin (because it's ugly) and support - as a filename indicating that stdin should be read?

That's already [supported](
https://github.com/sindresorhus/xo/blob/9a6513b72b201f2c7051850374127b3cee25db0c/cli.js#L145), but from experience, very few people know about it, and support for `-` in CLI apps is very inconsistent. I thought `--stdin` was clearer and more explicit.

**Comment by Qix-** [comment]
I disagree, I think `-` is the accepted way of indicating something be read from stdin.

`cat`:

> If file is a single dash (`-') or absent, cat reads from the standard input.

`tar`:

> The filename can be - for standard input or standard output.

`gzip`:

Undocumented, but

```console
$ echo hello | gzip - | gzip -d - | cat
hello
```

`vim`:

```console
$ echo hello | vim -
```

(opens a new vim document called `-stdin-` that has the buffer contents `hello` that can be `:saveas`'d)

`grep`:

```console
$ echo hello | grep -o "l" -
l
l
```

`awk`:

> The file name - means the standard input

```console
$ echo "hello world" | awk '{print $2}' -
world
```

I could go on. :dancers:

**Comment by sindresorhus** [comment]
Yeah. I wasn't arguing against Unix tools having `-`. Just that it's usually badly documented, so most wouldn't know about it, and most user-land tools doesn't support it, which definitely doesn't help the awareness of it. But yeah, I guess I could document `-` instead of `--stdin`. However, it doesn't really resolve the initial request here. So I assume there's no way to tell if stdin is available without actually trying to read it?

**Comment by Qix-** [comment]
Correct.

**Comment by Qix-** [comment]
> Just that it's usually badly documented, so most wouldn't know about it, and most user-land tools doesn't support it, which definitely doesn't help the awareness of it.

A few years later, I actually have a rebuttal to this.

`-` is actually specified in both the GNU argument parsing guidelines as well as the POSIX argument parsing guidelines. It's been around for decades.
